### PR TITLE
Enforce feature order

### LINKED ISSUE #264: Ordering of features and APIs
Currently, APIs and their sub-features are listed in arbitrary order. I think we should put them in the order they have been added to Firefox (then Chrome). There is also the suggestion to sort them alphabetically.

I prefer implementation order, since the story reads more clearly, having essential features that are widely supported listed first, followed by newer features. Also this seems consistent with the vast majority of existing documentation on MDN.

@wbamberg @elchi3

### PR DESCRIPTION
**Pull request by vinyldarkscratch**
This closes #264 and introduces the sorting of all the features by alphabetical order (case insensitive).  This also provides a script that automatically sorts said features (as well as the browser order in `__compat`), which can be called using `npm run fix`.

### CODE REVIEW

**Comment by Elchi3** [comment]
Again, the devil is in the details :) I think we need to talk a bit how we want to sort special characters. 

The characters allowed for the identifiers that we want to sort here are `a-zA-Z_0-9-$@` (see the schema file). I think should decide where to put things that start with a number, $, @, _ or -. I think with the sorting you propose here, these would be at the top. 
The "__compat" entry should always the first item if it is present and then I think I would rather have the rest at the bottom.

You can observe this for example on this table https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Browser_compatibility. There you have `@@match`, and so on, which are currently at the end, but normal sorting would put then at the start.

Another case is https://developer.mozilla.org/en-US/docs/Web/CSS/@media#Browser_compatibility where all the -moz- prefixed items etc would now be at the top when we right now have placed them at the end.

So, I've tried to change the sorting and here's what I came up with to put special chars last (change in fix-feature-order.js).

```js
value = Object.keys(value).sort((a, b) => {
      if(a === '__compat') return -1;
      return (b.toLowerCase().localeCompare("a") >= 0) - (a.toLowerCase().localeCompare("a") >= 0)
             || a.toLowerCase().localeCompare(b.toLowerCase());
    })
```


I didn't change it in the linter to see the difference to the current sorting, so here's the output of that and where this differs exactly from your original sorting. 

``` 
Problems in 24 files:
✖ api/CSSNumericArray.json
  Style – 1 error:
      Feature sorting error on line #53
    Actual:         "CSSNumericValue": {
    Expected:       "@@iterator": {
✖ api/CSSTransformValue.json
  Style – 1 error:
      Feature sorting error on line #53
    Actual:         "CSSTransformValue": {
    Expected:       "@@iterator": {
✖ api/CSSUnparsedValue.json
  Style – 1 error:
      Feature sorting error on line #53
    Actual:         "CSSTransformValue": {
    Expected:       "@@iterator": {
✖ api/HTMLCanvasElement.json
  Style – 1 error:
      Feature sorting error on line #156
    Actual:           "bitmaprenderer_context": {
    Expected:         "2d_alpha_context": {
✖ api/OffscreenCanvas.json
  Style – 1 error:
      Feature sorting error on line #133
    Actual:           "bitmaprenderer_context": {
    Expected:         "2d_context": {
✖ api/StylePropertyMapReadOnly.json
  Style – 1 error:
      Feature sorting error on line #53
    Actual:         "entries": {
    Expected:       "@@iterator": {
✖ css/at-rules/media.json
  Style – 1 error:
      Feature sorting error on line #55
    Actual:           "any-hover": {
    Expected:         "-moz-device-pixel-ratio": {
✖ css/properties/background-position.json
  Style – 1 error:
      Feature sorting error on line #54
    Actual:           "multiple_backgrounds": {
    Expected:         "4_value_syntax": {
✖ css/properties/background-repeat.json
  Style – 1 error:
      Feature sorting error on line #54
    Actual:           "multiple_backgrounds": {
    Expected:         "2-value": {
✖ css/properties/border-radius.json
  Style – 1 error:
      Feature sorting error on line #101
    Actual:           "elliptical_borders": {
    Expected:         "4_values_for_4_corners": {
✖ css/types/length-percentage.json
  Style – 1 error:
      Feature sorting error on line #55
    Actual:           "cap": {
    Expected:         "1in_is_96px": {
✖ css/types/length.json
  Style – 1 error:
      Feature sorting error on line #55
    Actual:           "cap": {
    Expected:         "1in_is_96px": {
✖ javascript/builtins/Array.json
  Style – 1 error:
      Feature sorting error on line #58
    Actual:           "concat": {
    Expected:         "@@iterator": {
✖ javascript/builtins/ArrayBuffer.json
  Style – 1 error:
      Feature sorting error on line #58
    Actual:           "byteLength": {
    Expected:         "@@species": {
✖ javascript/builtins/Date.json
  Style – 1 error:
      Feature sorting error on line #59
    Actual:           "getDate": {
    Expected:         "@@toPrimitive": {
✖ javascript/builtins/Map.json
  Style – 1 error:
      Feature sorting error on line #69
    Actual:           "clear": {
    Expected:         "@@iterator": {
✖ javascript/builtins/RegExp.json
  Style – 1 error:
      Feature sorting error on line #58
    Actual:           "compile": {
    Expected:         "@@match": {
✖ javascript/builtins/Set.json
  Style – 1 error:
      Feature sorting error on line #69
    Actual:           "add": {
    Expected:         "@@iterator": {
✖ javascript/builtins/String.json
  Style – 1 error:
      Feature sorting error on line #58
    Actual:           "anchor": {
    Expected:         "@@iterator": {
✖ javascript/builtins/Symbol.json
  Style – 1 error:
      Feature sorting error on line #59
    Actual:           "asyncIterator": {
    Expected:         "@@toPrimitive": {
✖ javascript/builtins/TypedArray.json
  Style – 1 error:
      Feature sorting error on line #58
    Actual:           "buffer": {
    Expected:         "@@iterator": {
✖ javascript/functions.json
  Style – 1 error:
      Feature sorting error on line #111
    Actual:           "callee": {
    Expected:         "@@iterator": {
✖ webextensions/api/devtools.json
  Style – 1 error:
      Feature sorting error on line #27
    Actual:               "inspect": {
    Expected:             "$0": {
✖ webextensions/manifest/commands.json
  Style – 1 error:
      Feature sorting error on line #25
    Actual:           "F1-F12": {
    Expected:         "_execute_sidebar_action": {
``` 

Long story short: I don't know how you or everyone else feels about this, but I think we should have special char things like the `@@match` etc. at the bottom.

**Comment by Elchi3** [comment]
@wbamberg since you've been involved in the sorting issue (#264), would you mind commenting on how to sort special characters? (see my suggestion above)

**Comment by vinyldarkscratch** [comment]
I'm in favor of putting them towards the bottom, especially if that's how we've got it currently sorted.  I went ahead and stashed those changes to `orderFeatures()` in both the fix and test scripts, so they're ready to go!

**Comment by wbamberg** [comment]
I think the main reason for sorting features is to help people find things, and give them the simplest, most consistent and most predictable rule for finding things.

So for me the answer to "how should we sort special characters?" is "how would most people expect special characters to be sorted?". Personally, I'd expect non-alpha characters to come first, just because that's the default in `sort()`, so it's a kind of lowest common denominator.

But I don't feel strongly about this. The main things are to have the check in the linter and the fix command.

I saw this is using `localeCompare()` - I don't know much about locale handling in JavaScript, but does this mean it might give different results in different environments?

**Comment by vinyldarkscratch** [comment]
Ah, unfortunately it looks like it may be giving different results.  I just gave it a quick test run by merging all of the sorting PRs into a new branch, along with the latest master, and it looks like there's over a thousand files that are being changed from the new code.

We should probably look further into what exactly `localeCompare()` does.

**Comment by Elchi3** [comment]
Thanks, Will!

> "how would most people expect special characters to be sorted?".

Yes, this question was driving my thinking as well. I believe for the two examples I've given, my view would be that people would expect special things like `@@match` or `-webkit-animation` would come last in the table, as everything else seems to be much more important and less obscure.

Given this seems to only affect sorting in 24 files, I think we could go with the "special characters last" approach for now, but change to standard "special characters first" sort, if it causes problems (I don't think it will).

> I saw this is using localeCompare() - I don't know much about locale handling in JavaScript, but does this mean it might give different results in different environments?

Good question. I think nodejs now supports internationalization, so this might matter.

**Comment by vinyldarkscratch** [comment]
A wonderful [StackOverflow answer](https://stackoverflow.com/a/52766466/2608838) gave me all the info I needed to come up with a sort function that sorts in the following order:

1. Letters (starting with uppercase)
2. Letters (starting with lowercase)
3. Numbers
4. Symbols

I also looked into `String.prototype.localeCompare()` a little more, and it appears that we can pass a locale in, therefore keeping results more consistent.  We can also disable sensitivity between accents, capitalization, punctuation, etc. if need be, though I don't think we'd need to configure any of that.

A commit will be coming soon to apply this new function!

**Comment by ddbeck** [comment]
@vinyldarkscratch it looks like you've merged a bunch of unreviewed PRs? Can you fill me on what has happened here?

**Comment by vinyldarkscratch** [comment]
Sure thing — the PRs were extensions off of their original PRs to correct a couple of sorting issues.  Since the original PRs were already approved and merged, I figured it would be easier (and would prevent merge conflicts from coming up between other PRs) to merge them as soon as Travis CI finished.  Should I refrain from doing as such, though?

**Comment by ddbeck** [comment]
Thanks for the update! And yeah, we'd prefer to avoid merging PRs without a review. Not a problem though—I'll push a commit to revert those merges and then we can go through the usual review process together. 🔀

**Comment by ddbeck** [comment]
@vinyldarkscratch I've reverted those changes with 2085ca9. You can see a complete list of PRs impacted in the commit message. If you'd open replacements, I'd be happy to review them. Thank you!

**Comment by chrisdavidmills** [comment]
I can also help review a couple if you like; I appreciate that reviewing these might take a little while.

**Comment by vinyldarkscratch** [comment]
Sounds like a plan, and thank you!  I restored the branches and will be opening PRs shortly to replace the existing ones.  😉

**Comment by vinyldarkscratch** [comment]
And all the PRs have now been restored!  If you don't mind me asking, I would love to prioritize them to prevent nasty merge conflicts that come up when new data is submitted, though beforehand (and this is where I think I got too hasty) we should determine how we wish to sort properties.  Does the order [proposed in my above comment](https://github.com/mdn/browser-compat-data/pull/3427#issuecomment-507097959) sound reasonable, @chrisdavidmills?

**Comment by chrisdavidmills** [comment]
@vinyldarkscratch order looks OK, except that in my experience numbers come before letters and symbols? This seems to be what the SO post is saying too.

**Comment by vinyldarkscratch** [comment]
We can sort numbers first, yeah!  At this point, I'd say it's mostly based upon what we think is the most practical for the data, and how it should be displayed in MDN.  I'll make the change to sort numbers first.  :wink:

**Comment by vinyldarkscratch** [comment]
And done, all sorting PRs have been updated to reflect this as well!

**Comment by ddbeck** [comment]
@vinyldarkscratch A heads up here, I'm not certain I'll have enough time to review this this week (I saw you tagged me in a review, I'm not ignoring it). If I can't get to this week, I'll give you some sort of update on Monday. Thank you!

**Comment by vinyldarkscratch** [comment]
Not a problem, thanks for taking the time to review it!  There’s definitely no rush.  😉

**Comment by ddbeck** [comment]
Hey @vinyldarkscratch, thanks for your patience on this. Would you be open to sort of rebooting this process, to try out some of the things we've discussed on #4469? I can draw up some next steps for that, if you're interested. Or if you'd rather press on with how things have started here, we can leave the #4469 to another issue. Let me know what you'd prefer.

**Comment by vinyldarkscratch** [comment]
We've been referring to this PR in the discussion in #4469.  Yeah, let's give this one a reboot with the new plans!  What would you like me to do to get this prepared for the new process?

**Comment by ddbeck** [comment]
OK, supposing we were starting from square one, here's how I'd imagine migrations would look:

* New migrations would go in a dedicated directory, probably `/scripts/migrations/` (e.g., move your fix-it script to `/scripts/migrations/sort-features.js`)
* A migration script itself would take some JSON files as arguments. For instance, I could run `node ./scripts/migrations/sort-features.js css/**` to fix all of CSS, or run it against a specific JSON file to see what happens. Hopefully this would change the files in place.
* A migration script would come with some test case(s) to demonstrate the behavior. I haven't given this a ton of thought. Something as simple as a test input file and an expected output file to diff against could work. But I'd also be OK with introducing something like Jest as a dev dependency, if that would make things more readable and maintainable—then you could have a tidy unit test for the compare function, for example.

As it pertains to this PR specifically, I'd expect:

* Other PRs related to this to close.
* If you want to include the linter in this PR, then I'd expect that the invocation would be commented out or gated in some way (e.g., setting environment variable). In other words, merging this PR shouldn't cause CI failures.

As you can tell, I'm making this up as I go, so I'm open to ideas or improvements on this.

**Comment by vinyldarkscratch** [comment]
Sounds like a plan to me!  Regarding the location of the script, personally, I feel that the fix-up script should remain where it is for two reasons:

1) We're expecting the sorting script to be something that's run more than just one during this bulk change.  As such, I feel it should be convenient and accessible by a quick `npm run lint`.
2) A refactor I plan to do is for both the fix script and its linter to import common functions and not use copy and paste.  Keeping them in the same folder should help with that process.

I'm not stubborn on this.  We can easily set up a fix script that uses the scripts within that folder!  Which do you think is more appropriate?

**Comment by ddbeck** [comment]
> We're expecting the sorting script to be something that's run more than just one during this bulk change

This is a great point and I didn't think of it. Yeah, leave the script where it is.
